### PR TITLE
Login: Add support for 2FA codes with white spaces

### PR DESCRIPTION
**Pull request by stephanethomas**
This pull request - along server-side patch D10671-code - fixes https://github.com/Automattic/wp-calypso/issues/21067 by updating the [`Verification Code` page](http://calypso.localhost:3000/log-in/sms) to accept codes formatted into chunks in order to improve their readability:

screenshot

#### Testing instructions

1. Run `git checkout update/login-2fa-codes` and start your server, or open a [live branch](https://calypso.live/?branch=update/login-2fa-codes)
2. See instructions in D10671-code

#### Reviews

- [ ] Code
- [x] Product

### CODE REVIEW

**Comment by p3ob7o** [comment]
It is impossible to type a space while typing verification code during login, which is confusing since that's how the code was received: xxxx xxx
There is code trimming spaces at the end of the input in real time.

It is possible to add the space after typing the code, in which case then the form validation refuses the input.

screen shot 2018-03-07 at 11 51 25

The form should accept input both with and without space.

Enhancement suggestion: What would be even less confusing would be to replace the space with a "-" and then split the form in two parts, with the "-" hardcoded in place, to remove any potential confusion.

**Comment by p3ob7o** [comment]
In the account recovery section it is possible to type the space, but then form validation silently refuses it by disabling the Validate button.
screen shot 2018-03-07 at 11 57 04
The button becomes active upon deletion of the space:
screen shot 2018-03-07 at 11 58 06

**Comment by stephanethomas** [comment]
> In the account recovery section it is possible to type the space, but then form validation silently refuses it by disabling the Validate button.

I've just fixed this issue.

**Comment by p3ob7o** [comment]
Both flows now work and accept the codes with and without spaces.

File: client/state/account-recovery/settings/actions.js [hunk 1]
```diff
@@ -229,7 +233,7 @@ export const validateAccountRecoveryPhone = code => dispatch => {
 	return wpcom
 		.undocumented()
 		.me()
-		.validateAccountRecoveryPhone( code )
+		.validateAccountRecoveryPhone( replace( code, ' ', '' ) )
```
Comment (yurynix) [inline client/state/account-recovery/settings/actions.js]: This replaces only first occurrence
Not sure why should we use that instead of `String.prototype.replace` ?

screen shot 2018-03-07 at 16 02 03

Comment (stephanethomas) [reply client/state/account-recovery/settings/actions.js]: > This replaces only first occurrence

Good catch, I'll fix that.

> Not sure why should we use that instead of String.prototype.replace ?

I've opted to use Lodash's `replace()` since it handles `undefined` and `null` values for `code` gracefully.

**Comment by lancewillett** [comment]
Hi team @yurynix @Tug @p3ob7o 

UX feedback sent in via @johnmaeda today in Slack — wanted to share with you.

- It looks odd to have the space after the 4th digit; why do we need the space now? (In the US it might look more logical to have it after the 3rd digit)
- Could the code just be for digits only like other services?

**Comment by p3ob7o** [comment]
I agree that XXX XXXX is more common than XXXX XXX, at least in terms of phone numbers.

The reason why that space was inserted is that XXXXXXX is hard to parse and memorize when going from the SMS app to the WordPress app.

My vote would be to reduce the length of the code to six digits like almost the entirety of other providers and then remove the space. Trying to change user habits in this type of situations is generally just bad UX.

**Comment by p3ob7o** [comment]
> Could the code just be for digits only like other services?

Did you mean **four** digits? That seems too weak, and below global standards of 2FA codes that are of six digits.

**Comment by lancewillett** [comment]
> My vote would be to reduce the length of the code to six digits like almost the entirety of other providers and then remove the space. Trying to change user habits in this type of situations is generally just bad UX.

This would be a good improvement. 👍

**Comment by lancewillett** [comment]
For history sake, here's the original change: https://github.com/Automattic/wp-calypso/issues/21067

**Comment by johnmaeda** [comment]
Love six digits with no spaces. That'd be a huge improvement (1 less character for a user to type). If we could do five, that would be even better. Some of my 2FAs use 4 (like Simple Bank by BBVA). But as for those who use 6 with *no space character break*:

* PayPal
* Github
* Facebook
* Google

Thank you @lancewillett @p3ob7o !

**Comment by p3ob7o** [comment]
I just tested the SMS code, and the last digit is actually useless: the first six are identical to the code generated by the authenticator app, and typing only the first six digits of the SMS code gets you logged in.

@yurynix could you see how to remove the extra digit from the SMS code?
